bin/metababel

```diff
@@ -87,30 +99,51 @@ end
 # Indeeed the <%= body %> will be indented, but we don't don't want it,
 # in the body string is already indented
 # But we clean the white space empty line afterward \o/
-def wrote_event_dispatchers(folder, t)
+def wrote_event_dispatchers(folder, t, m, callback_types)
   event_name = 'event'
+
+  matching_dispatchers = m.map do |md|
+    event, clock = t.find_event_class { |e| Regexp.new(md[:regex]).match?(e.name) }
+    raise "Matching callback #{md[:name]} does not match any event in model." unless event
+
+    args = {}
+    event.get_getter(event: '', arg_variables: args)
+    
+    dispatcher_args = []
+    dispatcher_args += [GeneratedArg.new('int64_t', '_timestamp')] if clock
+    dispatcher_args += args.fetch('outputs',[]).filter_map { |a| a if md[:args].include?(a.name) }
+    raise "Matching callback argument not present in model." unless dispatcher_args.length == md[:args].length
+
+    DispatcherInfo.new(md[:name], dispatcher_args, md[:name], Regexp.new(md[:regex]))
+  end 
+
   event_class_dispatchers = t.map_event_classes_with_index do |e, index_stream_class, index_event_class, default_clock_class|
     arg_variables = {}
     body = Babeltrace2Gen.context(indent: 1) do
       e.get_getter(event: event_name, arg_variables: arg_variables)
     end
-    EventClassInfo.new(e.name, arg_variables.fetch('outputs', []), body, index_stream_class, index_event_class, default_clock_class,
-                       arg_variables.fetch('outputs_allocated', []))
-  end
 
-  stream_classes_matching_dispatchers = t.stream_classes.map do |s|
-    arg_variables = {}
-    body = ''
-    if s.event_common_context_field_class
-      body = Babeltrace2Gen.context(indent: 1) do
-        s.get_getter(event: event_name, arg_variables: arg_variables)
-      end
-    end
-    StreamClassInfo.new(s.name, arg_variables.fetch('outputs', []), body, s.default_clock_class)
+    # Event dispatcher
+    evt_args = []
+    evt_args += [GeneratedArg.new('int64_t', '_timestamp')] if default_clock_class
+    evt_args += arg_variables.fetch('outputs',[])
+    dispatchers = [DispatcherInfo.new(e.name, evt_args, 'event')]
+
+    # Matching dispatchers
+    dispatchers += matching_dispatchers.filter_map { |md| md if md.regex.match?(e.name) }
+
+    EventClassInfo.new(
+      e.name, arg_variables.fetch('outputs', []), 
+      body, index_stream_class, index_event_class, 
+      default_clock_class,
+      arg_variables.fetch('outputs_allocated', []),
+      dispatchers
+    )
   end
 
   d = { event_class_dispatchers: event_class_dispatchers,
-        stream_classes_matching_dispatchers: stream_classes_matching_dispatchers,
+        matching_dispatchers: matching_dispatchers,
```

Comment:
Why do you still have 2 types of dispatchers?

Reply:
The main problem is that we need to render separate versions of `btx_register_callbacks_,`: `btx_register_matching_callback`  and `btx_register_event_callback`. The implementation of these functions are different. The former registers a callback to all the matched events. The later just register it to the associated event. So I need matching_dispatchers to render `btx_register_matching_callback` .